### PR TITLE
(wip) Refactor pause actions

### PR DESCRIPTION
**Pull request by jasonLaster**
Associated Issue: #4297 

### Summary of Changes

* cleanup actions
* defer expensive work
* remove ensure text

### Test Plan

more unit tests :)

### CODE REVIEW

**Comment by jasonLaster** [comment]
This change brings the time it takes to pause `perf.html` from 7s to 163ms. I kinda feel bad bragging about this because 7s is so large.

How did we get these savings? We deferred mapping the scopes until after we've saved the set the pause data. We're also not blocking on that data as well.

File: src/actions/sources/tests/createPrettySource.spec.js [hunk 1]
```diff
@@ -2,9 +2,11 @@ import { actions, createStore, makeSource } from "../../../utils/test-head";
 import { createPrettySource } from "../createPrettySource";
 import { getSourceByURL } from "../../../selectors";
 
+import { sourceThreadClient } from "../../tests/helpers/threadClient.js";
+
 describe("createPrettySource", () => {
-  it("returns a pretty source for a minified file", async () => {
-    const { dispatch, getState } = createStore();
+  fit("returns a pretty source for a minified file", async () => {
```
Comment (codehag) [inline src/actions/sources/tests/createPrettySource.spec.js]: forced test here

File: src/actions/pause/commands.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ * Debugger commands like stepOver, stepIn, stepUp
+ *
+ * @param string $0.type
+ * @memberof actions/pause
+ * @static
+ */
+export function command(type: CommandType) {
+  return async ({ dispatch, client }: ThunkArgs) => {
+    // execute debugger thread command e.g. stepIn, stepOver
+    dispatch({ type: "COMMAND", value: { type } });
```
Comment (codehag) [inline src/actions/pause/commands.js]: command data doesnt seem to be consumed anywhere? this might be replaced with an api call and then we can skip the redux cycle (since no front end data needs to be updated)

Comment (codehag) [inline src/actions/pause/commands.js]: is the data from COMMAND used anywhere?

File: src/actions/pause/paused.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Debugger has just paused
+ *
+ * @param {object} pauseInfo
+ * @memberof actions/pause
+ * @static
+ */
+export function paused(pauseInfo: Pause) {
+  return async function({ dispatch, getState, client, sourceMaps }: ThunkArgs) {
```
Comment (codehag) [inline src/actions/pause/paused.js]: do we need to use async here? or can we make this dispatch utilize thunk, and put the async (mapped frames, and frameScopes) stuff in the [PROMISE] ?

File: src/actions/pause/resumed.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * @memberof actions/pause
+ * @static
+ */
+export function resumed() {
+  return ({ dispatch, client, getState }: ThunkArgs) => {
+    if (!isPaused(getState())) {
+      return;
+    }
+
+    const wasPausedInEval = pausedInEval(getState());
```
Comment (codehag) [inline src/actions/pause/resumed.js]: perhaps we need a comment about this

File: src/actions/sources/loadSourceText.js [hunk 5]
```diff
@@ -1,8 +1,25 @@
 // @flow
 import { PROMISE } from "../../utils/redux/middleware/promise";
 import { setEmptyLines, setSymbols } from "../ast";
+import { getSource } from "../../selectors";
+import { setSource } from "../../workers/parser";
 import type { Source } from "../../types";
 import type { ThunkArgs } from "../types";
+
+async function loadSource(source: Source, { sourceMaps, client }) {
```
Comment (codehag) [inline src/actions/sources/loadSourceText.js]: 👍

File: src/actions/tests/sources.spec.js [hunk 6]
```diff
@@ -131,8 +133,8 @@ describe("sources", () => {
     const fooSource = getSource(getState(), "foobar.js");
     const barSource = getSource(getState(), "barfoo.js");
 
-    expect(fooSource.get("text").indexOf("return 2")).not.toBe(-1);
-    expect(barSource.get("text").indexOf("return 3")).not.toBe(-1);
+    expect(fooSource.get("text").indexOf("return foobar.js")).not.toBe(-1);
```
Comment (codehag) [inline src/actions/tests/sources.spec.js]: text is a string? if it is maybe `includes` would be clearer

File: src/components/SecondaryPanes/Scopes.js [hunk 7]
```diff
@@ -41,11 +41,12 @@ class Scopes extends PureComponent<Props, State> {
   }
 
   componentWillReceiveProps(nextProps) {
-    const { pauseInfo, selectedFrame } = this.props;
+    const { pauseInfo, selectedFrame, frameScopes } = this.props;
     const pauseInfoChanged = pauseInfo !== nextProps.pauseInfo;
-    const selectedFrameChange = selectedFrame !== nextProps.selectedFrame;
+    const selectedFrameChanged = selectedFrame !== nextProps.selectedFrame;
+    const frameScopesChanged = frameScopes !== nextProps.frameScopes;
```
Comment (codehag) [inline src/components/SecondaryPanes/Scopes.js]: is this part of the pause refactor?

File: src/components/shared/SearchInput.js [hunk 8]
```diff
@@ -6,7 +6,7 @@ import CloseButton from "./Button/Close";
 import "./SearchInput.css";
 
 const arrowBtn = (onClick, type, className, tooltip) => {
-  var props = {
+  const props = {
```
Comment (codehag) [inline src/components/shared/SearchInput.js]: :+1:

File: src/reducers/breakpoints.js [hunk 9]
```diff
@@ -185,24 +184,19 @@ export function getBreakpointForLine(
...
 }
 
-export const getHiddenBreakpoint = createSelector(getBreakpoints, function(
-  breakpoints
-) {
-  const hiddenBreakpoints = breakpoints
+export function getHiddenBreakpoint(state: OuterState) {
+  return getBreakpoints(state)
     .valueSeq()
     .filter(breakpoint => breakpoint.hidden)
```
Comment (codehag) [inline src/reducers/breakpoints.js]: maybe we could use `find`. it will return the first match, though it doesnt need to be done in this pr.. this also looks like it isnt part of the pause refactor. I bring this up mostly so that we have a clear chronology when looking back on the commit messages. don't worry about it for this one, but it might be worth splitting out these driveby refactors into their own prs

Comment (jasonLaster) [reply src/reducers/breakpoints.js]: yeah... i did it because `createSelector` was no longer defined... but that might be more troubling :/

File: src/actions/tests/sources.spec.js (hunk 6 above)
Comment (jasonLaster) [reply src/actions/tests/sources.spec.js]: agreed, but lets do this later :)
